### ISSUE TITLE
crypto: document which cipher.Block implementations are safe for concurrent use

### ISSUE THREAD

**Issue by FiloSottile** [golang/go#25882]
Many of them are, but since it's undocumented it might change from one architecture to the next.

This came up in https://go-review.googlesource.com/c/crypto/+/118535 because x/crypto/xts effectively assumes they all are.

**Comment by nodo** [comment]
Hey @FiloSottile ! Are you OK if I help with this issue? Do you have any pointers?

**Comment by FiloSottile** [comment]
@nodo Absolutely! It's a matter of going through the implementations (mainly AES and DES) and checking that they don't write to object state in Encrypt / Decrypt. You can also use the race detector to help, but some implementations are in assembly, so it will not work everywhere.

**Comment by nodo** [comment]
Hey @FiloSottile,

I was checking the packages, would it be enough to verify the files below?

- https://github.com/golang/go/blob/master/src/crypto/aes/gcm_amd64.s
- https://github.com/golang/go/blob/master/src/crypto/aes/block.go
- https://github.com/golang/go/blob/master/src/crypto/des/block.go

**Comment by FiloSottile** [comment]
There are a few other implementations for other platforms in those packages that you'll have to check. By the way, this is about cipher.Block, so you don't have to check the GCM mode, which is a cipher.AEAD implementation.

**Comment by anitgandhi** [comment]
I think the same should apply to some of the other interfaces and implementations. For example, `cipher.BlockMode`, and specifically the CBCEncrypter/Decrypter types, have `iv` as a shared state, so it's not concurrent safe. This is equivalent to the XTS tweak situation. 

I can open a separate issue for this if it makes more sense to do so.

**Comment by FiloSottile** [comment]
In general nothing is concurrent safe until explicitly marked. cipher.Block
was confusing because they mostly (all) are and that assumption had made
its way into places. If you think it would be worth explicitly documenting
other implementations, feel free to open another issue, though.
